### PR TITLE
Move to Java 11 for building and running

### PR DESCRIPTION
**Pull request by mcdan**
## Description
Move to using java 11 for building and running the containers.

## Related issue and scope

- [X] (Someone) opened an issue to propose and discuss this change (#4217)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [X] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [X] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [X] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [X] I reviewed the [style guides](https://github.com/apache/openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

File: tests/performance/gatling_tests/build.gradle [hunk 1]
```diff
@@ -26,6 +26,13 @@ repositories {
     mavenCentral()
 }
 
+sourceSets {
+    gatling {
+        scala.srcDirs = ["src/gatling/scala/"]
+        resources.srcDirs = ["src/gatling/resources/"]
```
Comment (markusthoemmes) [inline tests/performance/gatling_tests/build.gradle]: What would the default setup look like? Should we move our stuff to conform to the default setup?

File: tests/performance/gatling_tests/build.gradle [hunk 2]
```diff
@@ -16,7 +16,7 @@
  */
 
 plugins {
-    id "com.github.lkishalmi.gatling" version "0.7.2"
+    id "com.github.lkishalmi.gatling" version "3.2.1"
```
Comment (markusthoemmes) [inline tests/performance/gatling_tests/build.gradle]: whaaaaat. Nice! 🎉

File: tests/performance/gatling_tests/build.gradle (hunk 1 above)
Comment (mcdan) [reply tests/performance/gatling_tests/build.gradle]: We could move, yes, figured I'd land the update and make sure I didn't blow up the world then I can do another one to move us to their new layout. I will open a issue for that now.

**Comment by mcdan** [comment]
@dubee -
> Are all of the `JAVA_OPTS` in the Ansible deployment roles still valid with the newer version of Java?

I looked there and nothing seemed to out of the ordinary. Also doesn't travis use ansible to get a cluster up to test it?

> Also, the default garbage collector in Java 8 was Parallel GC, whereas the default in Java 11 is G1 GC. Has there been any investigation to determine if we should use G1 GC or Parallel GC?

No I haven't looked beyond the normal performance tests, do you have something in mind to run?

**Comment by dubee** [comment]
A performance test using each type of GC while comparing memory consumption of the controller and invoker should tell us.

**Comment by rabbah** [comment]
@mcdan can you resolve the conflicts?

File: common/scala/Dockerfile [hunk 3]
```diff
@@ -14,8 +14,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-
-FROM adoptopenjdk/openjdk8:x86_64-alpine-jdk8u172-b11
+FROM adoptopenjdk/openjdk11:x86_64-alpine-jdk-11.0.3_7
```
Comment (rabbah) [inline common/scala/Dockerfile]: We are using adoptopenjdk/openjdk11:x86_64-debian-jdk-11.0.5_10 in production with no issues.

**Comment by rabbah** [comment]
I sent a note to the dev list because I didn't see one https://lists.apache.org/thread.html/9acf4c4e213a4385c14f4bc9e4a1724ff79189ce4f35bf5ce1e239c6%40%3Cdev.openwhisk.apache.org%3E

I will merge this PR next week assuming conflicts are resolved.

**Comment by rabbah** [comment]
Should we replace this file also? https://github.com/apache/openwhisk/blob/master/tools/ubuntu-setup/java8.sh

**Comment by dgrove-oss** [comment]
@mhenke1 -- any update on the performance verification you had mentioned doing in https://lists.apache.org/thread.html/a2cb739c82a1f9e4e311c9d223fc1629f6c7514826b223b4b2948869%40%3Cdev.openwhisk.apache.org%3E

We should move forward on this PR.

**Comment by mhenke1** [comment]
Rodric,

performance is looking good. We did not find any additional problems.

But we are in the final stage of preparing a accompanying PR in our system. 
It got necessary to adapt some JVM setting we use to enhance the GC output.

That means that from our side we will be ok to merge the change in the beginning of next week.
I will give you an update on Monday.

Regards,
Martin

**Comment by rabbah** [comment]
Thank you Martin for checking the performance. Are the JVM settings which you noted relevant to the Apache project / would the project also benefit from the setting adjustments you're making?

**Comment by rabbah** [comment]
@mcdan I rebased this to master to remove the conflicts but it looks like I don't have permission to push to your fork. Can you do it then?

**Comment by dubee** [comment]
@rabbah, from what I've read, the default GC in Java 11 is optimized for low latency garbage collection. Whereas the old default for Java 8, Parallel GC, is optimized for throughput. 

There may be more discussion on this subject as we may have evidence of performance degradation using G1 GC.

**Comment by rabbah** [comment]
Thanks James for the additional details.

File: common/scala/Dockerfile (hunk 3 above)
Comment (rabbah) [inline common/scala/Dockerfile]: Per suggestion on the dev list - this reduces performance degradation. On the tech interchange call we discussed making this a separate PR but I think based on the dev list exchange we can just accept this commit and move on.
```suggestion
FROM adoptopenjdk/openjdk11-openj9:x86_64-alpine-jdk-11.0.5_10_openj9-0.17.0
```

**Comment by rabbah** [comment]
Per the discussion on the tech interchange call, we are accepting this PR as is. A subsequent PR can apply this suggested change https://github.com/apache/openwhisk/pull/4706#discussion_r381441881 per the dev list note from @mhenke1.

**Comment by mhenke1** [comment]
My proposed change can be found as an unmerged pull request to the originating branch of this PR:
https://github.com/adobe-apiplatform/incubator-openwhisk/pull/12

**Comment by rabbah** [comment]
Thanks Martin. Would you mind rebasing that branch? I can open the PR then.

**Comment by mhenke1** [comment]
Unfortunately my private OW Fork is utterly broken. I can not even delete it. I will open a PR as soon this is working again.

**Comment by rabbah** [comment]
yikes - that maybe explains why it is not showing up in the pull down list for creating a new PR.

**Pull request merged**